resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/builders/ParallelBuildChainTest.java

```diff
@@ -141,6 +141,17 @@ public void testIndividualProjectBuilds_ProjectRelaxedRule() throws Exception {
 		});
 	}
 
+	@Test
+	public void testIndividualProjectBuilds_WithManyProjects_ProjectRelaxedRule() throws Exception {
+		int numberOfParallelBuilds = 60;
```

Comment:
i increased it to 600 and got java.lang.AssertionError: not all build jobs have started in time
Expected: (a collection containing <P/longRunningBuildProject0> and a collection containing <P/longRunningBuildProject1> and a collection containing <P/longRunningBuildProject10> and a collection containing <P/longRunningBuildProject100> and a collection containing <P/longRunningBuildProject101> and a collection containing <P/longRunningBuildProject102> and a collection containing <P/longRunningBuildProject103> and a collection containing 
....

Reply:
i increased org.eclipse.core.tests.internal.builders.ParallelBuildChainTest.TIMEOUT_IN_MILLIS to not run in timeout and the threads are stuck in
```
"Worker-584: Building longRunningBuildProject85": BLOCKED
    org.eclipse.core.internal.jobs.LockManager.removeLockWaitThread(LockManager.java:287)
    org.eclipse.core.internal.jobs.ThreadJob.waitForRun(ThreadJob.java:350)
    org.eclipse.core.internal.jobs.ThreadJob.joinRun(ThreadJob.java:214)
    org.eclipse.core.internal.jobs.ImplicitJobs.begin(ImplicitJobs.java:100)
    org.eclipse.core.internal.jobs.JobManager.beginRule(JobManager.java:342)
    org.eclipse.core.internal.resources.WorkManager.checkIn(WorkManager.java:124)
    org.eclipse.core.internal.resources.Workspace.prepareOperation(Workspace.java:2403)
    org.eclipse.core.internal.resources.Project$1.run(Project.java:591)
    org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2453)
    org.eclipse.core.internal.resources.Project.internalBuild(Project.java:642)
    org.eclipse.core.internal.resources.Project.build(Project.java:154)
    org.eclipse.core.tests.internal.builders.ParallelBuildChainTest$2.run(ParallelBuildChainTest.java:440)
    org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)

```

Reply:
Thanks for the validation! I'll try to reproduce and figure out why this still fails.

Reply:
for numberOfParallelBuilds = 300 this PR drastically improves the test. 
sidenote: the wait logic was added with https://bugs.eclipse.org/bugs/show_bug.cgi?id=320329

Reply:
I finally found out the following: The threads are actually not stuck but only execution only becomes really slow. The reason is that all the jobs start build operations that first acquire the `WorkspaceRoot` rule for notifying about the build. The more jobs do that in parallel, the more complex the dependency graph becomes, which the deadlock detection has to process everytime one of the jobs is not able to acquire the lock. These calculations become very expensive.

You can easily see that progress is still made but hat it becomes very slow when increasing the number of jobs by logging whenever one job is allowed proceed here:

https://github.com/eclipse-platform/eclipse.platform/blob/ca35cee037256b0f6213f84acbf9de28a9650431/runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/ThreadJob.java#L287-L292

On my laptop, 400 jobs can be processed in around 60 seconds, while more jobs will take too long. 300 jobs take around 30 seconds and 200 jobs finish almost instant.

Note that this only happens in case `ThreadJob.waitForRun` is called for a huge number of jobs with conflicting rules at the same time. I am not sure whether this is a likely scenario in realistic use cases.

So in summary, the detected issue is "only" a performance issue, of which I am unsure whether it is worth (or even necessary) to be addressed. From my point of view, it is not related to the change proposed in this PR.

Reply:
how can a job releases the scheduling rule while it is running? shouldn' it be constant because it is set only before scheduling?

Reply:
Job can **transfer** rule, not sure if that is what you are asking for.

Reply:
i never did that which method is it?

Reply:
Actually I've mixed up a bit, but both could be interesting
- org.eclipse.core.runtime.jobs.IJobManager.transferRule(ISchedulingRule, Thread)
- org.eclipse.core.runtime.jobs.Job.yieldRule(IProgressMonitor)

Reply:
transferRule does not change jobs rule or state, 
yieldRule doesn't change the rule it changes state RUNNING -> YIELDING which can't happen while jobstate is synchronized, but yes could happen between manager.runNow(threadJob, true) and synchronize. This PR would fix that. Now i see a usecase why it can help

Reply:
It's a bit more tricky: while those places you mention may also lead to the discussed issue, the more relevent situations occur with respect to `ThreadJobs`, so the ones internally used by the `JobManager`. Those `ThreadJobs` are recycled and reused whenever such an internal job for the same thread is required, see:

https://github.com/eclipse-platform/eclipse.platform/blob/ca35cee037256b0f6213f84acbf9de28a9650431/runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/ThreadJob.java#L419

and

https://github.com/eclipse-platform/eclipse.platform/blob/ca35cee037256b0f6213f84acbf9de28a9650431/runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/ImplicitJobs.java#L201-L201

So whenever one job acquires multiple rules, one after another, using `JobManager::beginRule`, the same `ThreadJob` will be reused with different rules:
https://github.com/eclipse-platform/eclipse.platform/blob/ca35cee037256b0f6213f84acbf9de28a9650431/runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/JobManager.java#L340-L343

This is, in particular, the case for build operation, as performed in the failing test. They acquire three rules, the first and last being the `WorkspaceRoot` for notifications about build start/end (which is one leading to conflicts that affect the issue addressed by this PR), and the middle one being the actual build rule, see:

https://github.com/eclipse-platform/eclipse.platform/blob/ca35cee037256b0f6213f84acbf9de28a9650431/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Project.java#L576

Reply:
Thanks for explanation. Feels like would be better to not recycle ThreadJob. The sideeffects of recycling outweight any potential benefit.
Anyway the fix looks ok.

Reply:
Fully agree. I gave disabling the recycling functionality a short try and did not face any immediate regressions:

https://github.com/eclipse-platform/eclipse.platform/blob/ca35cee037256b0f6213f84acbf9de28a9650431/runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/ImplicitJobs.java#L225-L228

 In particular, the test we have used in this PR still runs fine and with the same performance when using a high number of jobs. I will create a follow-up issue and PR to investigate whether we can remove the recycling functionality to reduce complexity.

Thanks for the thorough review and validation, @jukzi!